design-proposals/klip-15-new-api-and-client.md

```diff
@@ -0,0 +1,187 @@
+* The protocol will follow tried and tested designs proved over many years by some of the most well known and best performing messaging
+ systems in the industry.
+* The protocol will be designed for generic flow controlled streaming of data, not limited to KSQL. Therefore it has potential for re-use in other
+Confluent projects which might have a need for high performance streaming. 
+
+### The server
+
+* We will create a new server implementation to host the new protocol.
+* The implementation will support equivalents for all the operations that are currently available on the HTTP/REST API
+* The implementation will be designed in a reactive / non blocking way in order to provide the best performance / scalability characteristics with
+low resource usage. This will also influence the overall threading model of the server and position us with a better, more scalability internal
+server architecture that will help future proof the ksqlDB server.
+* The problematic parts of the current HTTP API will be removed (query streaming using chunked response and websockets)
+* The remainder of the current HTTP API will remain supported as-is using the new implementation.
+
+## What is not in scope
+
+* We are not creating a new RESTful API.
+* We are not currently looking to write clients for languages other than Java and JavaScript (although those may follow later)
+* We will not initially allow for consuming directly from Kafka topics using the client (although tha may follow later)
+
+## Value/Return
+
+We hope that by providing a delightful, easy to use client, it will enable application developers to easily write powerful
+applications that take advantage of their data plane / event streaming platform more effectively.
+
+We hope that this could be transformative for the project in terms of adoption as it would position ksqlDB as a great choice for
+writing typical event sourcing / CQRS / DDD style applications, which are currently hard to write using ksqlDB alone. 
+
+There are also further incidental advantages gained by this work. By using a modern server side implementation such as Vert.x 
+there are benefits in relation to performance, scalability, simplicity of implementation, 
+and threading model. Not to mention reduced dependencies and better resource usage. This will set us up better for the kinds of high
+throughput operations that we will need to implement efficiently now that the project has pivoted to a more active application
+facing server rather than a more passive DDL engine.
+
+## Public APIS
+
+The following changes / additions to APIs will occur:
+
+* The current chunked streaming and websockets streaming endpoints will be retired.
+* The client will provide a new public API
+* We will provide a new wire protocol for client<->server communication.
+
+## Design
+
+### The client
+
+* The Java client will provide an API based on JDK 9 flow / reactive streams and completable future
```

Comment:
Aren't we going to have conflicts with other JDK versions used by CP? I think JDK 9 is not an officially supported version yet for CP, right?

Reply:
Yes, that is a problem. I really hope we can upgrade from Java 8 (Java 8 is EOL already!), but if not we can implement our own copy of the Java 9 Flow API, and then, when we do eventually upgrade to Java >= 9, we can implement the correct interfaces and deprecate the old ones.
We can also consider shipping a simple wrapper that wraps the client with actual Java 9 flow interfaces and is only distributed as part of the OSS project (so no problems with CP compatibility).
But I think a better solution would be to get CP moving on from Java 8, it can't stay there for ever!

Reply:
Or we just backport JDK 9 flow :)

Reply:
FYI we've already backported the flow API: https://github.com/confluentinc/ksql/blob/master/ksql-rest-app/src/main/java/io/confluent/ksql/rest/server/resources/streaming/Flow.java

Reply:
Cool! I did see that, but I think we would need to copy it to the client libs - we don't want to the user apps pulling in all the server side dependencies because our flow API lives in a server module.